File: aoc2022/day16/part2.py
import collections
import sys


class Memo:
    def __init__(self, valves, tunnels):
        self.valves = valves
        self.tunnels = tunnels
        self.memo = dict()

    def dfs(self, person, elephant, open_valves, mins_remaining):
        key = (min(person, elephant), max(person, elephant), tuple(open_valves), mins_remaining)

        if key in self.memo:
            return self.memo[key]

        result = self._dfs(person, elephant, open_valves, mins_remaining)
        self.memo[key] = result
        return result

    def _dfs(self, person, elephant, open_valves, mins_remaining):
        if mins_remaining <= 0:
            return 0

        current_rate = sum([self.valves[valve] for valve in open_valves])
        if len(open_valves) == len(self.tunnels) - 1:
            return current_rate * mins_remaining

        result = 0

        if person not in open_valves and self.valves[person] > 0:
            new_open_valves = {x for x in open_valves}
            new_open_valves.add(person)

            for elephant_dest in self.tunnels[elephant]:
                result = max(result, current_rate + self.dfs(person, elephant_dest, new_open_valves, mins_remaining - 1))

        if elephant not in open_valves and self.valves[elephant] > 0:
            new_open_valves = {x for x in open_valves}
            new_open_valves.add(elephant)

            for person_dest in self.tunnels[person]:
                result = max(result, current_rate + self.dfs(person_dest, elephant, new_open_valves, mins_remaining - 1))

        if person != elephant and person not in open_valves and elephant not in open_valves and self.valves[person] > 0 and self.valves[elephant] > 0:
            new_open_valves = {x for x in open_valves}
            new_open_valves.add(person)
            new_open_valves.add(elephant)
            result = max(result, current_rate + self.dfs(person, elephant, new_open_valves, mins_remaining - 1))

        for person_dest in self.tunnels[person]:
            for elephant_dest in self.tunnels[elephant]:
                result = max(result, current_rate + self.dfs(person_dest, elephant_dest, open_valves, mins_remaining - 1))

        return result


def get_min_distance(start, end, tunnels):
    queue = [(start, 0)]
    visited = set()

    while queue:
        curr, dist = queue[0]
        queue = queue[1:]

        if curr == end:
            return dist

        if curr in visited:
            continue
        visited.add(curr)

        for neighbor in tunnels[curr]:
            queue.append((neighbor, dist + 1))


def bypass_zero_nodes(valves, tunnels):
    result = collections.defaultdict(list)

    nodes = [k for k in valves.keys() if valves[k] > 0]
    nodes.append("AA")

    for i, start in enumerate(nodes):
        for end in nodes[i+1:]:
            if start == end:
                continue

            distance = get_min_distance(start, end, tunnels)

            result[start].append((end, distance))
            result[end].append((start, distance))
    return result


def run(filename):
    valves, tunnels = parse_file(filename)

    #tunnels = bypass_zero_nodes(valves, tunnels)

    memo = Memo(valves, tunnels)
    return memo.dfs("AA", "AA", set(), 26)


def parse_file(filename):
    valves = dict()
    tunnels = dict()

    with open(filename, 'r') as f:
        for line in f.readlines():
            tokens = line.split()

            node = tokens[1]

            rate = int(tokens[4][:-1].split("=")[1])

            paths = [tok.split(",")[0] for tok in tokens[9:]]

            valves[node] = rate
            tunnels[node] = paths

    return valves, tunnels


if __name__ == "__main__":
    print(run(sys.argv[1]))

File: aoc2022/day16/part2.rs
use std::cmp;
use std::collections::HashMap;
use std::collections::HashSet;
use std::env;
use std::fs;

fn main() {
    let args: Vec<String> = env::args().collect();
    let filename: &String = &args[1];

    let contents: String = fs::read_to_string(filename).expect("failed to read file");
    let lines: std::str::Lines = contents.lines();

    let mut valves: HashMap<&str, i64> = HashMap::new();
    let mut tunnels: HashMap<&str, Vec<&str>> = HashMap::new();

    for line in lines {
        let tokens: Vec<&str> = line.split_whitespace().collect();

        let node: &str = tokens[1];

        // e.g. rate=13;
        let rate: i64 = tokens[4].split(&['=', ';'][..]).nth(1).unwrap().parse::<i64>().unwrap();

        let paths: Vec<&str> = tokens[9..].iter().map(|token| token.split(',').next().unwrap()).collect();

        valves.insert(node, rate);
        tunnels.insert(node, paths);
    }

    let result: i64 = dfs(&mut HashMap::new(), "AA", "AA", &HashSet::new(), 26, &valves, &tunnels);

    println!("result: {}", result);
}

#[derive(PartialEq, Eq, Hash)]
struct MemoKey {
    val: String,
}

fn get_memo_key(person: &str, elephant: &str, open_valves: &HashSet<&str>, mins_remaining: i64) -> MemoKey {
    let mut v: Vec<&&str> = open_valves.into_iter().collect();
    v.sort();

    let mut result: String = cmp::max(person, elephant).to_string();
    result += &cmp::min(person, elephant).to_string();
    for val in v.iter() {
        result += val;
    }

    result += &mins_remaining.to_string();

    return MemoKey { val: result };
}

fn dfs(memo: &mut HashMap<MemoKey, i64>, person: &str, elephant: &str, open_valves: &HashSet<&str>, mins_remaining: i64,
       valves: &HashMap<&str, i64>, tunnels: &HashMap<&str, Vec<&str>>) -> i64 {
    let key: MemoKey = get_memo_key(person, elephant, open_valves, mins_remaining);
    if memo.contains_key(&key) {
        return *(memo.get(&key).unwrap());
    }

    if mins_remaining <= 0 {
        return 0;
    }

    let current_rate: i64 = open_valves.iter().map(|valve| valves.get(valve).unwrap()).sum();

    let mut result: i64 = 0;

    if !open_valves.contains(person) && valves.get(person).unwrap() > &0 {
        let mut open_valves_copy: HashSet<&str> = open_valves.clone();
        open_valves_copy.insert(person);

        for elephant_dest in tunnels.get(elephant).unwrap() {
            if mins_remaining == 26 {
                println!("recursing into {}, {}", person, elephant_dest);
            }
            result = cmp::max(result, current_rate + dfs(memo, person, elephant_dest, &open_valves_copy, mins_remaining - 1, valves, tunnels));
        }
    }
    if !open_valves.contains(elephant) && valves.get(elephant).unwrap() > &0 {
        let mut open_valves_copy: HashSet<&str> = open_valves.clone();
        open_valves_copy.insert(elephant);

        for person_dest in tunnels.get(person).unwrap() {
            if mins_remaining == 26 {
                println!("recursing into {}, {}", person_dest, elephant);
            }
            result = cmp::max(result, current_rate + dfs(memo, person_dest, elephant, &open_valves_copy, mins_remaining - 1, valves, tunnels));
        }
    }
    if person != elephant && !open_valves.contains(person) && !open_valves.contains(elephant) {
        let mut open_valves_copy: HashSet<&str> = open_valves.clone();
        open_valves_copy.insert(person);
        open_valves_copy.insert(elephant);

        result = cmp::max(result, current_rate + dfs(memo, person, elephant, &open_valves_copy, mins_remaining - 1, valves, tunnels));
    }

    for person_dest in tunnels.get(person).unwrap() {
        for elephant_dest in tunnels.get(elephant).unwrap() {
            if mins_remaining == 26 {
                println!("recursing into {}, {}", person_dest, elephant_dest);
            }
            result = cmp::max(result, current_rate + dfs(memo, person_dest, elephant_dest, open_valves, mins_remaining - 1, valves, tunnels));
        }
    }

    memo.insert(key, result);
    //println!("found partial result: ({}, {}, {:?}, {}) = {}", person, elephant, open_valves, mins_remaining, result);
    return result;
}
